anomalib/models/cfa/lightning_model.py

```diff
@@ -0,0 +1,140 @@
+"""Lightning Implementatation of the CFA Model.
+
+CFA: Coupled-hypersphere-based Feature Adaptation for Target-Oriented Anomaly Localization
+
+Paper https://arxiv.org/abs/2206.04325
+"""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from typing import List, Optional, Tuple, Union
+
+import torch
+from omegaconf import DictConfig, ListConfig
+from pytorch_lightning import Callback
+from pytorch_lightning.callbacks import EarlyStopping
+from pytorch_lightning.utilities.cli import MODEL_REGISTRY
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from torch import Tensor
+from torch.optim.optimizer import Optimizer
+
+from anomalib.models.cfa.torch_model import CfaModel
+from anomalib.models.components import AnomalyModule
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["Cfa", "CfaLightning"]
+
+
+@MODEL_REGISTRY
+class Cfa(AnomalyModule):
+    """CFA: Coupled-hypersphere-based Feature Adaptation for Target-Oriented Anomaly Localization.
+
+    Args:
+        input_size (Tuple[int, int]): Size of the model input.
+        backbone (str): Backbone CNN network
+        gamma_c (int, optional): gamma_c value from the paper. Defaults to 1.
+        gamma_d (int, optional): gamma_d value from the paper. Defaults to 1.
+    """
+
+    def __init__(self, input_size: Tuple[int, int], backbone: str, gamma_c: int = 1, gamma_d: int = 1) -> None:
+        super().__init__()
+        self.model: CfaModel = CfaModel(input_size=input_size, backbone=backbone, gamma_c=gamma_c, gamma_d=gamma_d)
+
+    def on_train_start(self) -> None:
+        """Initialize the centroid for the memory bank computation."""
+        self.model.initialize_centroid(data_loader=self.trainer.datamodule.train_dataloader())  # type: ignore
+
+    def training_step(self, batch) -> STEP_OUTPUT:
+        """Training step for the CFA model.
+
+        Args:
+            batch (dict): Batch input.
+
+        Returns:
+            STEP_OUTPUT: Loss value.
+        """
+        loss = self.model(batch["image"])
+        return {"loss": loss}
+
+    def validation_step(self, batch, batch_idx) -> dict:
+        """Validation step for the CFA model.
+
+        Args:
+            batch (dict): Input batch.
+            batch_idx (int): Index of the batch.
+
+        Returns:
+            dict: Anomaly map computed by the model.
+        """
+        batch["anomaly_maps"] = self.model(batch["image"])
+        return batch
+
+    # pylint: disable=unused-argument
+    def backward(
+        self, loss: Tensor, optimizer: Optional[Optimizer], optimizer_idx: Optional[int], *args, **kwargs
+    ) -> None:
+        """Backward step for the CFA model.
+
+        Args:
+            loss (Tensor): Loss value.
+            optimizer (Optional[Optimizer]): Optimizer.
+            optimizer_idx (Optional[int]): Optimizer index.
+        """
+        # TODO: Investigate why retain_graph is needed.
+        loss.backward(retain_graph=True)
```

Comment:
Is this really needed? Did you observe different performance when just using the default `backward` method?

Reply:
For some reason, the training does not work without this. We could investigate this in the future.